pkgs/top-level/config.nix

```diff
@@ -0,0 +1,250 @@
+# This file defines the structure of the `config` nixpkgs option.
+
+{ lib, localSystem, crossSystem, config, ... }:
+
+with lib;
+
+let
+
+  mkMeta = args: mkOption (builtins.removeAttrs args [ "feature" ] // {
+    type = types.uniq (args.type or types.bool);
+    default = args.default or false;
+    description = args.description or ''
+      Whether to ${args.feature} while evaluating nixpkgs.
+    '' + ''
+      Changing the default will not cause any rebuilds.
+    '';
+  });
+
+  mkMassRebuild = args: mkOption (builtins.removeAttrs args [ "feature" ] // {
+    type = types.uniq (args.type or types.bool);
+    default = args.default or false;
+    description = (args.description or ''
+      Whether to ${args.feature} while building nixpkgs packages.
+    '') + ''
+      Changing the default will cause a mass rebuild.
+    '';
+  });
+
+  mkUse = args: mkMassRebuild (builtins.removeAttrs args [ "extra" ] // {
+    description = (args.description or ''
+      Whether to enable ${args.feature} support for nixpkgs packages.
+    '');
+  }) // {
+    extra = args.extra or (x: {});
+  };
+
+  # Flipped optionalAttrs variants for convenience
+  whenEnabled = args: value: lib.optionalAttrs value args;
+  whenDisabled = args: value: lib.optionalAttrs (!value) args;
+
+  use = config.use;
+
+in rec {
+
+  options = {
+
+    /* Internal stuff */
+
+    warnings = mkOption {
+      type = types.listOf types.str;
+      default = [];
+      internal = true;
+    };
+
+    /* Config options */
+
+    allowUnfree = mkMeta {
+      description = "Whether to allow evaluation of unfree packages.";
+    };
+
+    checkMeta = mkMeta {
+      feature = "check <literal>meta</literal> attributes of all the packages";
+    };
+
+    doCheckByDefault = mkMassRebuild {
+      feature = "run <literal>checkPhase</literal> by default";
+    };
+
+    /* callPackage scope */
+
+    extraScope = mkOption {
+      type = types.attrs;
```

Comment:
should be function type, similar to `packageOverrides`.
The example below doesn't work...

... which leads us to https://github.com/NixOS/nixpkgs/pull/44601 and added controversies. However here we require `types.attrSelector`